### ISSUE TITLE
CFP: Gateway API - support upstream http protocol selection based on appProtocol

### ISSUE THREAD

**Issue by rauanmayemir** [cilium/cilium#30452]
## Cilium Feature Proposal

Currently, `CiliumEnvoyConfig` is set up to `useDownstreamProtocolConfig`:

```
    - '@type': type.googleapis.com/envoy.config.cluster.v3.Cluster
      connectTimeout: 5s
      name: backend-ns/backend-service:http
      outlierDetection:
        splitExternalLocalOriginErrors: true
      type: EDS
      typedExtensionProtocolOptions:
        envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
          '@type': >-
            type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
          commonHttpProtocolOptions:
            idleTimeout: 60s
          useDownstreamProtocolConfig:
            http2ProtocolOptions: {}

```

It means that upstream backends have to support both protocols and there is no way to specify the protocol for it.

I'm suggesting to change this behavior to this:

1. Assume HTTP1.1 by default. I think it's a pretty sane assumption as most of common services, in my experience, expose just an HTTP endpoint with an option to set TLS (where it's also a sane assumption that TLS is usually terminated at the ingress).

```
envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
  '@type': >-
    type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
  commonHttpProtocolOptions:
    idleTimeout: 60s
  explicitHttpConfig:
    httpProtocolOptions: {}
```

2. Apply HTTP2 if the backendRef service specifies HTTP2/H2C:

```
apiVersion: v1
kind: Service
metadata:
  name: backend-service
  namespace: backend-ns
spec:
  selector:
    app.kubernetes.io/name: MyApp
  ports:
    - name: http
      appProtocol: kubernetes.io/h2c 
      protocol: TCP
      port: 80
      targetPort: 8080
```

```
envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
  '@type': >-
    type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
  commonHttpProtocolOptions:
    idleTimeout: 60s
  explicitHttpConfig:
    http2ProtocolOptions: {}
```

---

I understand that dropping `useDownstreamProtocolConfig` altogether might break all the existing setups, so I'm not sure how to deal with it without making it too opinionated. One thing that comes to mind is using port name conventions like istio does it. I.e if port name is prefixed with `http` - it's likely a regular HTTP1.1 endpoint. if it specifies something among `https`, `grpc`, `tls`, then it's probably safe to send both HTTP1.1 and HTTP2 to this endpoint.

**Comment by rauanmayemir** [comment]
Cc: @youngnick gentle nudge.
I'd love to get your feedback and what might not work here. I'm very interested in the subject.

**Comment by lmb** [comment]
Have you considered joining the weekly developer meeting?

**Comment by rauanmayemir** [comment]
@lmb I'll join the next one. 🙌

**Comment by AnthonyPoschen** [comment]
i am trying to explicitly get my service serving with HTTP2 and not sure how / where to do it ( go backend ). I also couldn't find anything in the docs talking about the default behaviour and how to control it so something like what is proposed here would at least be very clear on how to do it, although id just prefer if it could be HTTP2 in the tls handshake it just did it.

**Comment by rauanmayemir** [comment]
Closing this as PR has been merged and went into 1.16.x.
